File: 10-code-control/13-flag-targeting-rules/go/main.go
// Console grid navigator demonstrating LaunchDarkly targeting rules.
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
	ld "github.com/launchdarkly/go-server-sdk/v7"
	"golang.org/x/term"
)

// LaunchDarkly targeting rules inspect the public team context attribute.
// No team intentionally omits the attribute so evaluation reaches fallthrough.
// https://launchdarkly.com/docs/home/flags/target-rules
const (
	flagTeamStyle = "configure-team-label-style"
	appBanner     = "13-flag-targeting-rules[go]"
	bgANSI        = "\033[48;5;236m"
	resetANSI     = "\033[0m"
)

var (
	rows     = [3]string{"t", "m", "b"}
	cols     = [3]string{"l", "m", "r"}
	ldClient *ld.LDClient
)

type login struct {
	username string
	team     string
}

type flagValues struct {
	teamLabel string
	style     string
}

type position struct {
	row, col int
}

type moveResult struct {
	row, col int
	moved    bool
}

type sessionAction int

const (
	actionQuit sessionAction = iota
	actionLogout
)

func initLaunchDarkly() {
	sdkKey := os.Getenv("LD_SDK_KEY")
	if sdkKey == "" {
		fmt.Fprintln(os.Stderr, "Warning: LD_SDK_KEY not set — flag uses plain default.")
		return
	}
	client, err := ld.MakeClient(sdkKey, 5*time.Second)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Warning: LaunchDarkly SDK did not initialize — flag uses plain default.")
		return
	}
	ldClient = client
}

// Evaluate the string variation with team as a public context attribute.
// https://launchdarkly.com/docs/home/flags/context-attributes
func evaluateFlags(username, team string) flagValues {
	builder := ldcontext.NewBuilder(username)
	if team != "" {
		builder.SetString("team", team)
	}
	context := builder.Build()

	style := "plain"
	if ldClient != nil {
		candidate, err := ldClient.StringVariation(flagTeamStyle, context, "plain")
		if err == nil {
			switch candidate {
			case "plain", "colored-red", "colored-blue", "colored-yellow":
				style = candidate
			}
		}
	}
	return flagValues{teamLabel: teamLabel(team), style: style}
}

func teamLabel(team string) string {
	switch team {
	case "red":
		return "Team Red"
	case "blue":
		return "Team Blue"
	case "yellow":
		return "Team Yellow"
	default:
		return "No team"
	}
}

func coloredTeam(flags flagValues) string {
	color := ""
	switch flags.style {
	case "colored-red":
		color = "\033[31m"
	case "colored-blue":
		color = "\033[34m"
	case "colored-yellow":
		color = "\033[33m"
	}
	if color == "" {
		return flags.teamLabel
	}
	return color + flags.teamLabel + resetANSI + bgANSI
}

func formatPos(row, col int) string {
	return rows[row] + "/" + cols[col]
}

func tryMove(row, col, dr, dc int) moveResult {
	nr := clamp(row+dr, 0, 2)
	nc := clamp(col+dc, 0, 2)
	return moveResult{nr, nc, nr != row || nc != col}
}

func clamp(value, low, high int) int {
	if value < low {
		return low
	}
	if value > high {
		return high
	}
	return value
}

// Prompt for the user key and public team attribute used by targeting rules.
func readLogin(reader *bufio.Reader) (login, error) {
	fmt.Println(appBanner)
	fmt.Println("Login")
	fmt.Println()
	var username string
	for username == "" {
		fmt.Print("Username: ")
		line, err := reader.ReadString('\n')
		if err != nil {
			return login{}, err
		}
		username = strings.TrimSpace(line)
		if username == "" {
			fmt.Println("Username is required.")
		}
	}

	teams := map[string]string{"1": "", "2": "red", "3": "blue", "4": "yellow"}
	for {
		fmt.Print("Team [1=None 2=Red 3=Blue 4=Yellow]: ")
		line, err := reader.ReadString('\n')
		if err != nil {
			return login{}, err
		}
		if team, ok := teams[strings.TrimSpace(line)]; ok {
			return login{username: username, team: team}, nil
		}
		fmt.Println("Choose 1, 2, 3, or 4.")
	}
}

func drawCell(selected bool, line int) string {
	if selected {
		return [3]string{"┏━━━┓", "┃ X ┃", "┗━━━┛"}[line]
	}
	return [3]string{"┌───┐", "│   │", "└───┘"}[line]
}

func writeLine(out *strings.Builder, line string) {
	out.WriteString(line)
	out.WriteString("\033[K\r\n")
}

func render(session login, row, col int, previous *position, flags flagValues) {
	var out strings.Builder
	out.WriteString(bgANSI)
	out.WriteString("\033[H\033[2J")

	previousText := "—"
	if previous != nil {
		previousText = formatPos(previous.row, previous.col)
	}
	writeLine(&out, appBanner)
	writeLine(&out, "Name: "+session.username)
	writeLine(&out, "Team: "+coloredTeam(flags))
	writeLine(&out, "Current position: "+formatPos(row, col))
	writeLine(&out, "Previous position: "+previousText)
	writeLine(&out, "")
	writeLine(&out, "Use arrow keys or WASD to move (L to logout, Q to quit).")
	writeLine(&out, "")

	for r := 0; r < 3; r++ {
		for line := 0; line < 3; line++ {
			cells := make([]string, 3)
			for c := 0; c < 3; c++ {
				cells[c] = drawCell(r == row && c == col, line)
			}
			writeLine(&out, strings.Join(cells, " "))
		}
	}
	fmt.Print(out.String())
}

func readKeyWithTimeout(reader *bufio.Reader, timeout time.Duration) (int, int, sessionAction, bool, bool) {
	if err := os.Stdin.SetReadDeadline(time.Now().Add(timeout)); err != nil {
		return 0, 0, actionQuit, false, true
	}
	defer os.Stdin.SetReadDeadline(time.Time{})

	key, err := reader.ReadByte()
	if err != nil {
		if timeoutError, ok := err.(interface{ Timeout() bool }); ok && timeoutError.Timeout() {
			return 0, 0, actionQuit, false, false
		}
		return 0, 0, actionQuit, true, true
	}
	switch key {
	case 3, 'q', 'Q':
		return 0, 0, actionQuit, true, true
	case 'l', 'L':
		return 0, 0, actionLogout, true, true
	case 'w', 'W':
		return -1, 0, actionQuit, false, true
	case 's', 'S':
		return 1, 0, actionQuit, false, true
	case 'a', 'A':
		return 0, -1, actionQuit, false, true
	case 'd', 'D':
		return 0, 1, actionQuit, false, true
	case 27:
		second, err := reader.ReadByte()
		if err != nil || second != '[' {
			return 0, 0, actionQuit, false, false
		}
		third, err := reader.ReadByte()
		if err != nil {
			return 0, 0, actionQuit, false, false
		}
		switch third {
		case 'A':
			return -1, 0, actionQuit, false, true
		case 'B':
			return 1, 0, actionQuit, false, true
		case 'C':
			return 0, 1, actionQuit, false, true
		case 'D':
			return 0, -1, actionQuit, false, true
		}
	}
	return 0, 0, actionQuit, false, false
}

// Re-evaluate the team style every 500 ms while navigating.
func runGrid(session login, reader *bufio.Reader) sessionAction {
	row, col := 1, 1
	var previous *position
	for {
		flags := evaluateFlags(session.username, session.team)
		render(session, row, col, previous, flags)
		dr, dc, action, endSession, hasMove := readKeyWithTimeout(reader, 500*time.Millisecond)
		if endSession {
			return action
		}
		if !hasMove {
			continue
		}
		result := tryMove(row, col, dr, dc)
		if result.moved {
			old := position{row, col}
			previous = &old
			row, col = result.row, result.col
		}
	}
}

func main() {
	initLaunchDarkly()
	defer func() {
		if ldClient != nil {
			ldClient.Close()
		}
		fmt.Print(resetANSI)
	}()

	reader := bufio.NewReader(os.Stdin)
	fd := int(os.Stdin.Fd())
	for {
		session, err := readLogin(reader)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		oldState, err := term.MakeRaw(fd)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		action := runGrid(session, reader)
		_ = term.Restore(fd, oldState)
		fmt.Print(resetANSI)
		if action == actionQuit {
			return
		}
	}
}

File: 10-code-control/13-flag-targeting-rules/rust/src/main.rs
//! Console grid navigator demonstrating LaunchDarkly targeting rules.

use crossterm::{
    cursor::{Hide, MoveTo, Show},
    event::{self, Event, KeyCode, KeyEvent, KeyModifiers},
    execute, queue,
    style::Print,
    terminal::{self, ClearType},
};
use launchdarkly_server_sdk::{AttributeValue, Client, ConfigBuilder, ContextBuilder};
use std::io::{self, Write};
use std::sync::Arc;
use std::time::Duration;

// LaunchDarkly targeting rules inspect the public team context attribute.
// No team omits it so evaluation reaches the plain fallthrough.
// https://launchdarkly.com/docs/home/flags/target-rules
const FLAG_TEAM_STYLE: &str = "configure-team-label-style";
const APP_BANNER: &str = "13-flag-targeting-rules[rust]";
const BG: &str = "\x1b[48;5;236m";
const RESET: &str = "\x1b[0m";
const ROWS: [&str; 3] = ["t", "m", "b"];
const COLS: [&str; 3] = ["l", "m", "r"];

struct Login {
    username: String,
    team: String,
}

struct FlagValues {
    team_label: String,
    style: String,
}

struct Position {
    row: i32,
    col: i32,
}

#[derive(PartialEq, Eq)]
enum SessionAction {
    Quit,
    Logout,
}

struct App {
    client: Option<Arc<Client>>,
}

impl App {
    fn new() -> Self {
        let sdk_key = std::env::var("LD_SDK_KEY").ok();
        let client = sdk_key.and_then(|key| {
            let config = ConfigBuilder::new(&key).build().ok()?;
            let client = Client::build(config).ok()?;
            client.start_with_runtime().ok()?;
            let deadline = std::time::Instant::now() + Duration::from_secs(5);
            while !client.initialized() && std::time::Instant::now() < deadline {
                std::thread::sleep(Duration::from_millis(50));
            }
            if client.initialized() {
                Some(Arc::new(client))
            } else {
                eprintln!(
                    "Warning: LaunchDarkly SDK did not initialize — flag uses plain default."
                );
                None
            }
        });
        if std::env::var("LD_SDK_KEY").is_err() {
            eprintln!("Warning: LD_SDK_KEY not set — flag uses plain default.");
        }
        Self { client }
    }

    /// Evaluate the string variation with team as a public context attribute.
    /// https://launchdarkly.com/docs/home/flags/context-attributes
    fn evaluate_flags(&self, username: &str, team: &str) -> FlagValues {
        let mut builder = ContextBuilder::new(username);
        if !team.is_empty() {
            builder.set_value("team", AttributeValue::String(team.to_string().into()));
        }
        let context = builder
            .build()
            .unwrap_or_else(|_| ContextBuilder::new("anonymous").build().unwrap());
        let candidate = self.client.as_ref().map_or_else(
            || "plain".to_string(),
            |client| client.str_variation(&context, FLAG_TEAM_STYLE, "plain".to_string()),
        );
        let style = match candidate.as_str() {
            "plain" | "colored-red" | "colored-blue" | "colored-yellow" => candidate,
            _ => "plain".to_string(),
        };
        FlagValues {
            team_label: team_label(team).to_string(),
            style,
        }
    }
}

impl Drop for App {
    fn drop(&mut self) {
        if let Some(client) = self.client.take() {
            if let Ok(client) = Arc::try_unwrap(client) {
                client.close();
            }
        }
    }
}

fn team_label(team: &str) -> &'static str {
    match team {
        "red" => "Team Red",
        "blue" => "Team Blue",
        "yellow" => "Team Yellow",
        _ => "No team",
    }
}

fn colored_team(flags: &FlagValues) -> String {
    let color = match flags.style.as_str() {
        "colored-red" => "\x1b[31m",
        "colored-blue" => "\x1b[34m",
        "colored-yellow" => "\x1b[33m",
        _ => "",
    };
    if color.is_empty() {
        flags.team_label.clone()
    } else {
        format!("{color}{}{RESET}{BG}", flags.team_label)
    }
}

/// Prompt for the user key and public team attribute used by targeting rules.
fn read_login() -> io::Result<Login> {
    println!("{APP_BANNER}");
    println!("Login\n");
    let username = loop {
        print!("Username: ");
        io::stdout().flush()?;
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        let name = line.trim().to_string();
        if !name.is_empty() {
            break name;
        }
        println!("Username is required.");
    };

    let team = loop {
        print!("Team [1=None 2=Red 3=Blue 4=Yellow]: ");
        io::stdout().flush()?;
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        match line.trim() {
            "1" => break "",
            "2" => break "red",
            "3" => break "blue",
            "4" => break "yellow",
            _ => println!("Choose 1, 2, 3, or 4."),
        }
    };
    Ok(Login {
        username,
        team: team.to_string(),
    })
}

fn format_pos(row: i32, col: i32) -> String {
    format!("{}/{}", ROWS[row as usize], COLS[col as usize])
}

fn cell_line(selected: bool, line: usize) -> &'static str {
    if selected {
        ["┏━━━┓", "┃ X ┃", "┗━━━┛"][line]
    } else {
        ["┌───┐", "│   │", "└───┘"][line]
    }
}

fn print_line(out: &mut impl Write, y: u16, text: &str) -> io::Result<()> {
    queue!(out, MoveTo(0, y), Print(text), Print("\x1b[K"))?;
    Ok(())
}

fn render(
    out: &mut impl Write,
    login: &Login,
    row: i32,
    col: i32,
    previous: Option<&Position>,
    flags: &FlagValues,
) -> io::Result<()> {
    execute!(
        out,
        MoveTo(0, 0),
        terminal::Clear(ClearType::All),
        Print(BG)
    )?;
    let previous_text = previous
        .map(|position| format_pos(position.row, position.col))
        .unwrap_or_else(|| "—".to_string());
    let lines = [
        APP_BANNER.to_string(),
        format!("Name: {}", login.username),
        format!("Team: {}", colored_team(flags)),
        format!("Current position: {}", format_pos(row, col)),
        format!("Previous position: {previous_text}"),
        String::new(),
        "Use arrow keys or WASD to move (L to logout, Q to quit).".to_string(),
        String::new(),
    ];
    let mut y = 0;
    for line in lines {
        print_line(out, y, &line)?;
        y += 1;
    }
    for r in 0..3 {
        for line in 0..3 {
            let row_text = (0..3)
                .map(|c| cell_line(r == row && c == col, line))
                .collect::<Vec<_>>()
                .join(" ");
            print_line(out, y, &row_text)?;
            y += 1;
        }
    }
    out.flush()
}

/// Re-evaluate the team style every 500 ms while navigating.
fn run_grid(out: &mut impl Write, app: &App, login: &Login) -> io::Result<SessionAction> {
    let (mut row, mut col) = (1, 1);
    let mut previous = None;
    loop {
        let flags = app.evaluate_flags(&login.username, &login.team);
        render(out, login, row, col, previous.as_ref(), &flags)?;
        if !event::poll(Duration::from_millis(500))? {
            continue;
        }
        if let Event::Key(KeyEvent {
            code, modifiers, ..
        }) = event::read()?
        {
            if modifiers.contains(KeyModifiers::CONTROL) {
                return Ok(SessionAction::Quit);
            }
            let movement = match code {
                KeyCode::Char('q' | 'Q') => return Ok(SessionAction::Quit),
                KeyCode::Char('l' | 'L') => return Ok(SessionAction::Logout),
                KeyCode::Up | KeyCode::Char('w' | 'W') => Some((-1, 0)),
                KeyCode::Down | KeyCode::Char('s' | 'S') => Some((1, 0)),
                KeyCode::Left | KeyCode::Char('a' | 'A') => Some((0, -1)),
                KeyCode::Right | KeyCode::Char('d' | 'D') => Some((0, 1)),
                _ => None,
            };
            if let Some((dr, dc)) = movement {
                let next_row = (row + dr).clamp(0, 2);
                let next_col = (col + dc).clamp(0, 2);
                if next_row != row || next_col != col {
                    previous = Some(Position { row, col });
                    row = next_row;
                    col = next_col;
                }
            }
        }
    }
}

fn main() -> io::Result<()> {
    let app = App::new();
    let mut stdout = io::stdout();
    loop {
        let login = read_login()?;
        terminal::enable_raw_mode()?;
        execute!(stdout, Hide)?;
        let action = run_grid(&mut stdout, &app, &login);
        execute!(stdout, Show, Print(RESET))?;
        terminal::disable_raw_mode()?;
        if action? == SessionAction::Quit {
            break;
        }
    }
    Ok(())
}
